### PR TITLE
#21900: Fallback to local DNS when no other nameservers are known

### PR DESCRIPTION
**Pull request by rl1987**
https://trac.torproject.org/projects/tor/ticket/21900

### CODE REVIEW

File: src/feature/relay/dns.c [hunk 1]
```diff
@@ -1385,42 +1385,63 @@ configure_nameservers(int force)
...
+      missing_resolv_conf = 1;
+    } else if (!force && resolv_conf_fname &&
+               !strcmp(conf_fname,resolv_conf_fname)
         && st.st_mtime == resolv_conf_mtime) {
-      log_info(LD_EXIT, "No change to '%s'", conf_fname);
+      log_info(LD_EXIT, "no change to '%s'", conf_fname);
       return 0;
     }
+
+    if (st.st_size == 0)
```
Comment (nmathewson) [inline src/feature/relay/dns.c]: If the stat() call failed above, then at this point, st.st_size will be uninitialized, and we're not allowed to look at it.

File: src/feature/relay/dns.c [hunk 2]
```diff
@@ -1385,42 +1385,63 @@ configure_nameservers(int force)
...
-               conf_fname, conf_fname, r);
-      goto err;
-    }
-    if (evdns_base_count_nameservers(the_evdns_base) == 0) {
-      log_warn(LD_EXIT, "Unable to find any nameservers in '%s'.", conf_fname);
-      goto err;
+#endif /* defined(dns_option_hostsfile) && defined(use_libseccomp) */
+
+    if (!missing_resolv_conf) {
+      log_info(LD_EXIT, "parsing resolver configuration in '%s'", conf_fname);
```
Comment (nmathewson) [inline src/feature/relay/dns.c]: It looks like this log message, and the messages after it, are still lower-cased in the latest version of the branch?

File: src/feature/relay/dns.c [hunk 3]
```diff
@@ -1385,42 +1385,63 @@ configure_nameservers(int force)
...
+                    // fallback.
+      }
+      if (evdns_base_count_nameservers(the_evdns_base) == 0) {
+        log_warn(LD_EXIT, "unable to find any nameservers in '%s'.",
+                 conf_fname);
+      }
+      resolv_conf_fname = tor_strdup(conf_fname);
+      resolv_conf_mtime = st.st_mtime;
+    } else {
+      log_warn(LD_EXIT, "Could not read your resolv.conf - "
```
Comment (nmathewson) [inline src/feature/relay/dns.c]: This message should say what actual filename it tried to read.

File: src/feature/relay/dns.c [hunk 4]
```diff
@@ -1385,42 +1385,63 @@ configure_nameservers(int force)
...
+                        "please investigate your DNS configuration. "
+                        "This is possibly a problem. Meanwhile, falling"
+                        " back to local DNS at 127.0.0.1.");
+      evdns_base_nameserver_ip_add(the_evdns_base, "127.0.0.1");
     }
+
     tor_free(resolv_conf_fname);
```
Comment (nmathewson) [inline src/feature/relay/dns.c]: This tor_free() here will clear the value of rsolv_conf_fname that you set above...

File: src/feature/relay/dns.c [hunk 5]
```diff
@@ -1385,42 +1385,63 @@ configure_nameservers(int force)
...
+
+        if (r != 6) // "r = 6" means "no DNS servers were in resolv.conf" -
+          goto err; // in which case we expect libevent to add 127.0.0.1 as
+                    // fallback.
+      }
+      if (evdns_base_count_nameservers(the_evdns_base) == 0) {
+        log_warn(LD_EXIT, "unable to find any nameservers in '%s'.",
+                 conf_fname);
+      }
+      resolv_conf_fname = tor_strdup(conf_fname);
```
Comment (nmathewson) [inline src/feature/relay/dns.c]: If resolv_conf_fname was already set, this will leak memory

File: src/feature/relay/dns.c (hunk 1 above)
Comment (rl1987) [reply src/feature/relay/dns.c]: Checking if `stat()` succeeded in 770aee4856e9a4a1ee8b338c6ba083117a86ad78.

File: src/feature/relay/dns.c (hunk 3 above)
Comment (rl1987) [reply src/feature/relay/dns.c]: Logging filename in a1656c9f84ad2e3e6aabd2aedf54264c44a455df.

File: src/feature/relay/dns.c (hunk 4 above)
Comment (rl1987) [reply src/feature/relay/dns.c]: Fixed - 4c1913a492e7f593b9abb46a1368635e8347496c.

File: src/feature/relay/dns.c (hunk 5 above)
Comment (rl1987) [reply src/feature/relay/dns.c]: 4c1913a492e7f593b9abb46a1368635e8347496c

**Comment by nmathewson** [comment]
Closing in favor of PR 425

**Pull request closed without merging**